Source/JavaScriptCore/heap/MarkedBlock.cpp

```diff
@@ -513,156 +553,105 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
-            if (*mem)
-                break;
-            ++mem;
-        }
-    }
-    LOG_INVALID_HANDLE_DETAILS("found %zd 0s at beginning of block\n", contiguousZeroCountAfterHandle);
-    SAVE_TO_REG(savedMarkedBlock, this);
-    SAVE_TO_REG(savedHeapCell, heapCell);
-    SAVE_TO_REG(savedType, cellType);
-    DEFINE_SAVED_VALUE(savedCount, "x22", contiguousZeroCountAfterHandle);
-
-    bool isValidBlockVM = false;
-    bool foundBlockInThisVM = false;
-    bool isBlockInVM = false;
-    bool isBlockHandleInVM = false;
-
     VM* blockVM = header().m_vm;
     VM* actualVM = nullptr;
-    DEFINE_SAVED_VALUE(savedBlockVM, "x23", blockVM);
-    DEFINE_SAVED_VALUE(savedActualVM, "x24", actualVM);
-    DEFINE_SAVED_VALUE(savedBitfield, "x25", 0L);
+    bool isBlockVMValid = false;
+    bool isBlockInSet = false;
+    bool isBlockInDirectory = false;
+    bool foundInBlockVM = false;
+    size_t contiguousZeroBytesHeadOfBlock = 0;
+    size_t totalZeroBytesInBlock = 0;
+    uint64_t cellFirst8Bytes = 0;
+
+    if (heapCell) {
+        uint64_t* p = bitwise_cast<uint64_t*>(heapCell);
+        cellFirst8Bytes = *p;
+    }
 
-    {
-        VMInspector::forEachVM([&](VM& vm) {
-            if (blockVM == &vm) {
-                isValidBlockVM = true;
-                SAVE_TO_REG(savedActualVM, &vm);
-                SAVE_TO_REG(savedBitfield, 8);
-                LOG_INVALID_HANDLE_DETAILS("block VM %p is valid\n", &vm);
-                return IterationStatus::Done;
-            }
-            return IterationStatus::Continue;
-        });
+    auto updateCrashLogMsg = [&](int line) {
+#if PLATFORM(COCOA)
+        StringPrintStream out;
+        out.printf("INVALID HANDLE [%d]: markedBlock=%p; heapCell=%p; cellFirst8Bytes=%#llx; contiguousZeros=%lu; totalZeros=%lu; blockVM=%p; actualVM=%p; isBlockVMValid=%d; isBlockInSet=%d; isBlockInDir=%d; foundInBlockVM=%d;",
+            line, this, heapCell, cellFirst8Bytes, contiguousZeroBytesHeadOfBlock, totalZeroBytesInBlock, blockVM, actualVM, isBlockVMValid, isBlockInSet, isBlockInDirectory, foundInBlockVM);
+        const char* msg = out.toCString().data();
+        WTF::setCrashLogMessage(msg);
+        dataLogLn(msg);
+#else
+    UNUSED_PARAM(line);
+#endif
+    };
+    updateCrashLogMsg(__LINE__);
+
+    char* blockStart = bitwise_cast<char*>(this);
+    bool sawNonZero = false;
+    for (auto mem = blockStart; mem < blockStart + MarkedBlock::blockSize; mem++) {
+        // Exclude the MarkedBlock::Header::m_lock from the zero scan since taking the lock writes a non-zero value.
+        auto isMLockBytes = [blockStart](char* p) ALWAYS_INLINE_LAMBDA {
+            constexpr size_t lockOffset = offsetOfHeader + OBJECT_OFFSETOF(MarkedBlock::Header, m_lock);
+            size_t offset = p - blockStart;
+            return lockOffset <= offset && offset < lockOffset + sizeof(MarkedBlock::Header::m_lock);
+        };
+        bool byteIsZero = !*mem;
+        if (byteIsZero || isMLockBytes(mem)) {
```

Comment:
Shouldn't this be:
```
    if (!isNotLockByte(men))
        continue;
    else if (byteIsZero) {
```
Otherwise, it looks like you're counting all non-lock bytes as 0 regardless of their value.

Reply:
Please see above. `isMLockBytes` should return true iff `offset` falls on `m_lock`.